emscripten.py

```diff
@@ -648,14 +648,17 @@ def include_asm_consts(pre, forwarded_json, metadata, settings):
 
   asm_consts, all_sigs = all_asm_consts(metadata)
   asm_const_funcs = []
-  for sig in set(all_sigs):
-    forwarded_json['Functions']['libraryFunctions']['_emscripten_asm_const_' + sig] = 1
+  for call_type, sig in set(all_sigs):
+    forwarded_json['Functions']['libraryFunctions']['_emscripten_asm_const_' + call_type + sig] = 1
     args = ['a%d' % i for i in range(len(sig)-1)]
     all_args = ['code'] + args
+    if shared.Settings.USE_PTHREADS and call_type == 'sync_on_main_thread_': proxy_to_main_thread = '  if (ENVIRONMENT_IS_PTHREAD) { ' + ('Runtime.warnOnce("proxying function " + code); ' if shared.Settings.PTHREADS_DEBUG else '') + 'return _emscripten_sync_run_in_browser_thread_' + sig + '(-1 - ' + ','.join(all_args) + '); } \n'
```

Comment:
These lines are pretty long. Can they be broken up?
There's a lot of repetition as well, where things only differ slightly (e.g., `Runtime.warnOnce(message)`), some intermediate variables would probably go a long way here.

Reply:
What on earth is `'(-1 - ' + ','.join(all_args) + ');` doing? Call this but with the first argument (the function index iirc) plus one, negated? Why?
My guess is it's a sign to defer the call to the main thread, undo the transform, and call the index there, but if so that is *magic*, and not in a good way.

Reply:
> What on earth is '(-1 - ' + ','.join(all_args) + '); doing?

The proxying mechanism here is the same as the proxying mechanism for regular compiled C function pointers, so the concept here is that positive ordinals `1, 2, 3, ...` are regarded as pointers to compiled code functions, and to reuse the same mechanism, negative ordinals `-1, -2, -3, ...` are used to refer to EM_ASM blocks. So this code is remapping the sequence `0, 1, 2, ...` over to `-1, -2, -3, ...`. I think I had a comment somewhere about this, but adding one here as well.